rust/src/http2/http2.rs

```diff
@@ -0,0 +1,1175 @@
+                            });
+                        }
+                    }
+                    Err(_) => {
+                        self.set_event(HTTP2Event::InvalidFrameData);
+                        return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                            reason: HTTP2FrameUnhandledReason::ParsingError,
+                        });
+                    }
+                }
+            }
+            Some(parser::HTTP2FrameType::HEADERS) => {
+                let dyn_headers = if dir == STREAM_TOCLIENT {
+                    &mut self.dynamic_headers_tc
+                } else {
+                    &mut self.dynamic_headers_ts
+                };
+                match parser::http2_parse_frame_headers(input, hflags, dyn_headers) {
+                    Ok((hrem, hs)) => {
+                        for i in 0..hs.blocks.len() {
+                            if hs.blocks[i].error
+                                >= parser::HTTP2HeaderDecodeStatus::HTTP2HeaderDecodeError
+                            {
+                                self.set_event(HTTP2Event::InvalidHeader);
+                            }
+                        }
+                        if hrem.len() > 0 {
+                            SCLogNotice!("Remaining data for HTTP2 headers");
+                            self.set_event(HTTP2Event::ExtraHeaderData);
+                        }
+                        return HTTP2FrameTypeData::HEADERS(hs);
+                    }
+                    Err(nom::Err::Incomplete(_)) => {
+                        if complete {
+                            self.set_event(HTTP2Event::InvalidFrameData);
+                            return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                                reason: HTTP2FrameUnhandledReason::ParsingError,
+                            });
+                        } else {
+                            return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                                reason: HTTP2FrameUnhandledReason::TooLong,
+                            });
+                        }
+                    }
+                    Err(_) => {
+                        self.set_event(HTTP2Event::InvalidFrameData);
+                        return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                            reason: HTTP2FrameUnhandledReason::ParsingError,
+                        });
+                    }
+                }
+            }
+            Some(parser::HTTP2FrameType::PING) => {
+                return HTTP2FrameTypeData::PING;
+            }
+            _ => {
+                return HTTP2FrameTypeData::UNHANDLED(HTTP2FrameUnhandled {
+                    reason: HTTP2FrameUnhandledReason::UnknownType,
+                });
+            }
+        }
+    }
+
+    fn stream_data(&mut self, dir: u8, input: &[u8], over: bool, txid: u64) {
+        //TODOask use streaming buffer directly
```

Comment:
Should we consider every content data a file on HTTP2 ?

On HTTP1, we do it only for responses from the server, not data posted by the client

Reply:
The HTTP1 handling is different for historical reasons. I think we can consider everything a file in http2, but we need to be mindful of things like compression (gzip etc), chunking, multi-part. Not sure if any of these apply to http2, but for http1 consider the normalized data to be the file.

Reply:
ok, so far, so good